### PR TITLE
Fix duplicate genesis blocks on testnet

### PR DESCRIPTION
**Pull request by KentShikama**
## Overview
<sup>_What this PR does, and why it's needed_</sup>

Adds logic in casper packet handler to save and restore approved block and implements save/restore in the block store. If we have an approved block already saved, then we don't generate a new one when the standalone node is restarted.

### JIRA ticket:
<sup>_Create it if there isn't one already._</sup>

https://rchain.atlassian.net/browse/RCHAIN-3089
https://rchain.atlassian.net/browse/RCHAIN-3090

### Notes
<sup>_Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else._</sup>

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by KentShikama** [comment]
bors try

**Comment by itegulov** [comment]
bors try

**Comment by KentShikama** [comment]
bors try

**Comment by KentShikama** [comment]
bors try-

**Comment by KentShikama** [comment]
bors try

**Comment by KentShikama** [comment]
bors try-

**Comment by KentShikama** [comment]
bors try

File: casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala [hunk 1]
```diff
@@ -270,13 +270,13 @@ object CasperPacketHandler extends CasperPacketHandlerInstances {
                      capserHandlerInternal
                    )
                  case Some(approvedBlock) =>
-                   val blockMessage = approvedBlock.candidate.flatMap(_.block).get
+                   val genesis = approvedBlock.candidate.flatMap(_.block).get
```
Comment (tymm) [inline casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala]: :+1:

File: block-storage/src/main/scala/coop/rchain/blockstorage/FileLMDBIndexBlockStore.scala [hunk 2]
```diff
@@ -182,6 +183,24 @@ class FileLMDBIndexBlockStore[F[_]: Monad: Sync: RaiseIOError: Log] private (
...
+          Some(ApprovedBlock.parseFrom(bytes))
+      }
+    )
+
+  def putApprovedBlock(block: ApprovedBlock): F[Unit] =
+    lock.withPermit {
+      val tmpFile = approvedBlockPath.resolveSibling(approvedBlockPath.getFileName + ".tmp")
+      writeToFile(tmpFile, block.toByteArray) >>
+        moveFile(tmpFile, approvedBlockPath, StandardCopyOption.ATOMIC_MOVE) >>
+        ().pure[F]
```
Comment (dzajkowski) [inline block-storage/src/main/scala/coop/rchain/blockstorage/FileLMDBIndexBlockStore.scala]: `.as(())`?

File: casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala [hunk 3]
```diff
@@ -42,101 +49,127 @@ object CasperPacketHandler extends CasperPacketHandlerInstances {
...
-    } else {
-      for {
-        validators  <- CasperConf.parseValidatorsFile[F](conf.knownValidatorsFile)
-        validatorId <- ValidatorIdentity.fromConfig[F](conf)
-        bootstrap <- Ref.of[F, CasperPacketHandlerInternal[F]](
-                      new BootstrapCasperHandler(
+    BlockStore[F].getApprovedBlock.flatMap {
+      case Some(approvedBlock) =>
+        // We have an approved block so the network is already up and so no genesis ceremony needed.
+        val genesis = approvedBlock.candidate.flatMap(_.block).get
```
Comment (tymm) [inline casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala]: But here `approvedBlock` not necessarily has to be the genesis block, right?

File: casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala [hunk 4]
```diff
@@ -590,6 +590,18 @@ object CasperPacketHandler extends CasperPacketHandlerInstances {
...
+  /*
+   * Note the ordering of the insertions is important.
+   * We always want the block dag store to be a subset of the block store.
+   */
+  private def insertIntoBlockAndDagStore[F[_]: Sync: Concurrent: ErrorHandler: TransportLayer: ConnectionsCell: Log: BlockStore: BlockDagStorage](
+      genesis: BlockMessage
+  ): F[Unit] =
+    for {
+      _ <- BlockStore[F].put(genesis.blockHash, genesis)
+      _ <- BlockDagStorage[F].insert(genesis, false)
```
Comment (dzajkowski) [inline casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala]: nitpick: name boolean parameter?

Comment (KentShikama) [reply casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala]:
```suggestion
      _ <- BlockDagStorage[F].insert(genesis, invalid = false)
```

**Comment by KentShikama** [comment]
bors try

File: casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala [hunk 5]
```diff
@@ -42,101 +49,127 @@ object CasperPacketHandler extends CasperPacketHandlerInstances {
...
+                     genesis,
+                     conf.shardId
+                   )
+          _                   <- MultiParentCasperRef[F].set(casper)
+          _                   <- Log[F].info("Making a transition to ApprovedBlockRecievedHandler state.")
+          abh                 = new ApprovedBlockReceivedHandler[F](casper, approvedBlock)
+          validator           <- Ref.of[F, CasperPacketHandlerInternal[F]](abh)
+          casperPacketHandler = new CasperPacketHandlerImpl[F](validator)
+        } yield casperPacketHandler
+      case None =>
```
Comment (dzajkowski) [inline casper/src/main/scala/coop/rchain/casper/util/comm/CasperPacketHandler.scala]: don't you feel this deserves a split into helper methods?

**Pull request merged**